src/ripple/app/misc/impl/AMMUtils.cpp

```diff
@@ -188,4 +188,110 @@ ammAccountHolds(
     return STAmount{issue};
 }
 
+static std::pair<TER, AllNodesDeleted>
+deleteAMMTrustLines(
+    ApplyView& view,
+    AccountID const& ammAccountID,
+    std::uint16_t maxTrustlinesToDelete,
+    beast::Journal j)
+{
+    return cleanupOnAccountDelete(
+        view,
+        keylet::ownerDir(ammAccountID),
+        [&](LedgerEntryType nodeType,
+            uint256 const&,
+            std::shared_ptr<SLE>& sleItem) -> TER {
+            // Should only have the trustlines
+            if (nodeType != LedgerEntryType::ltRIPPLE_STATE)
+            {
+                JLOG(j.error())
+                    << "deleteAMMTrustLines: deleting non-trustline "
+                    << nodeType;
+                return tecINTERNAL;
+            }
+
+            // Trustlines must have zero balance
+            if (sleItem->getFieldAmount(sfBalance) != beast::zero)
+            {
+                JLOG(j.error())
+                    << "deleteAMMTrustLines: deleting trustline with "
+                       "non-zero balance.";
+                return tecINTERNAL;
+            }
+
+            return deleteAMMTrustLine(view, sleItem, ammAccountID, j);
+        },
+        j,
+        maxTrustlinesToDelete);
+}
+
+TER
+deleteAMMAccount(
+    ApplyView& view,
+    Issue const& asset,
+    Issue const& asset2,
+    beast::Journal j)
+{
+    auto ammSle = view.peek(keylet::amm(asset, asset2));
+    if (!ammSle)
+        return tecINTERNAL;
+
+    auto const ammAccountID = (*ammSle)[sfAccount];
+    auto sleAMMRoot = view.peek(keylet::account(ammAccountID));
+    if (!sleAMMRoot)
```

Comment:
> Counterpoint: If the account root doesn't exist, something has gone very, very wrong. Continuing with the deletion process _may_ make things worse. Arguably, the ledger and transaction history should be examined manually to determine what happened and what's going on. The solution may require a fix amendment not just to clean up the ledger, but to prevent the issue from happening again.

This would be an ungraceful failure mode. Half an AMM, if it happened, would probably crash the ledger or block an orderbook. You can't clean it up because of this pedantic check, even after restarting the whole network. Now you need to wait for an amendment and code update. You should seek to build graceful failure modes into production systems.

Reply:
If the ledger is corrupt, we're already in a failure mode. Better to stop than to continue in an state that may not be correct.

https://xrpl.org/intro-to-consensus.html#consensus-protocol-properties
> These properties are sometimes summarized as the following principles, in order of priority: Correctness, Agreement, Forward Progress.

If correctness is lost, then we do not want to make forward progress. This is not just me being stubborn, this is baked into the design of the ledger from the beginning.